sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -0,0 +1,622 @@
+  // are not otherwise represented in the payload.
+  // (may force runners to execute the ParDo differently)
+  repeated Parameter parameters = 3;
+
+  // (Optional) An ordered list of side inputs, describing for each local name
+  // to the data to be provided and the expected access pattern.
+  // (the SDK may not be order-sensitive)
+  repeated SideInput side_inputs = 4;
+}
+
+// A coder, the binary format for serialization and deserialization of data in
+// a pipeline.
+message Coder {
+
+  // (Required) A cross-language, stable, unique identifier for the (possibly
+  // parametric) encoding.
+  string urn = 1;
+
+  // (Optional) If this coder is parametric, such as ListCoder(VarIntCoder),
+  // this is a list of the components. In order for encodings to be identical,
+  // the URN and all components must be identical, recursively.
+  repeated string component_coder_ref = 2;
+
+  // (Optional) The pipeline-scoped id for the FunctionSpec of an SDK-specific
+  // UDF implementing the encoding.
+  string custom_coder_fn_ref = 3;
+}
+
+// A windowing strategy describes the window function, triggering, allowed
+// lateness, and accumulation mode for a PCollection.
+//
+// TODO: consider inlining field on PCollection
+message WindowingStrategy {
+
+  // (Required) The pipeline-scoped id for the FunctionSpec of the UDF that
+  // assigns windows, merges windows, and shifts timestamps before they are
+  // combined according to the OutputTime.
+  string window_fn_ref = 1;
+
+  // (Required) Whether or not the window fn is merging.
+  //
+  // This knowledge is required for many optimizations.
+  MergeStatus merge_status = 2;
+
+  // (Required) The coder for the windows of this PCollection.
+  string window_coder_ref = 3;
+
+  // (Required) The trigger to use when grouping this PCollection.
+  Trigger trigger = 4;
+
+  // (Required) The accumulation mode indicates whether new panes are a full
+  // replacement for prior panes or whether they are deltas to be combined
+  // with other panes (the combine should correspond to whatever the upstream
+  // grouping transform is).
+  AccumulationMode accumulation_mode = 5;
+
+  // (Required) The OutputTime specifies, for a grouping transform, how to
+  // compute the aggregate timestamp. The window_fn will first possibly shift
+  // it later, then the OutputTime takes the max, min, or ignores it and takes
+  // the end of window.
+  //
+  // This is actually only for input to grouping transforms, but since they
+  // may be introduced in runner-specific ways, it is carried along with the
+  // windowing strategy.
+  OutputTime output_time = 6;
+
+  // (Required) The duration, in milliseconds, beyond the end of a window at
+  // which the window becomes droppable.
+  int64 allowed_lateness = 7;
+}
+
+// Whether or not a PCollection's WindowFn is non-merging, merging, or
+// merging-but-already-merged, in which case a subsequent GroupByKey is almost
+// always going to do something the user does not want
```

Comment:
Acknowledged. I think that setting this bit will be preferable to replacing the `WindowFn` with `InvalidWindows`, really.